pkg/services/datasources/service/datasource.go

```diff
@@ -252,6 +257,182 @@ func (s *Service) AddDataSource(ctx context.Context, cmd *datasources.AddDataSou
 	})
 }
 
+// This will valid validate the instance settings and mutate the cmd with the processed values
+func (s *Service) prepareAdd(ctx context.Context, cmd *datasources.AddDataSourceCommand) error {
+	operation := backend.InstanceSettingsOperationCREATE
+
+	if len(cmd.Name) > maxDatasourceNameLen {
+		return datasources.ErrDataSourceNameInvalid.Errorf("max length is %d", maxDatasourceNameLen)
+	}
+
+	if len(cmd.URL) > maxDatasourceUrlLen {
+		return datasources.ErrDataSourceURLInvalid.Errorf("max length is %d", maxDatasourceUrlLen)
+	}
+
+	if cmd.Type == "" {
+		return nil // This happens in tests
+	}
+
+	// Make sure it is a known plugin type
+	p, found := s.pluginStore.Plugin(ctx, cmd.Type)
+	if !found {
+		return fmt.Errorf("plugin '%s' not found", cmd.Type)
+	}
+
+	// When the APIVersion is set, the client must also implement ProcessInstanceSettings
+	if p.APIVersion == "" {
+		if cmd.APIVersion != "" {
+			return fmt.Errorf("invalid request apiVersion (datasource does not have one configured)")
+		}
+		return nil
+	}
+
+	jd, err := cmd.JsonData.ToDB()
+	if err != nil {
+		return fmt.Errorf("invalid jsonData (%v)", operation)
+	}
+
+	rsp, err := s.pluginClient.ProcessInstanceSettings(ctx, &backend.ProcessInstanceSettingsRequest{
+		PluginContext: backend.PluginContext{
```

Comment:
tell me more!  maybe

Reply:
https://github.com/grafana/grafana/blob/main/pkg/services/pluginsintegration/plugincontext/plugincontext.go

Reply:
😢  was afraid it would be circular
```

wire: /Users/ryan/workspace/grafana/grafana/pkg/server/wire.go:177:20: cycle for *github.com/grafana/grafana/pkg/services/pluginsintegration/plugincontext.Provider:
	*github.com/grafana/grafana/pkg/services/pluginsintegration/plugincontext.Provider (github.com/grafana/grafana/pkg/services/pluginsintegration/plugincontext.ProvideService) ->
	github.com/grafana/grafana/pkg/services/datasources.DataSourceService (github.com/grafana/grafana/pkg/services/datasources/service.ProvideService) ->
	*github.com/grafana/grafana/pkg/services/pluginsintegration/plugincontext.Provider
wire: /Users/ryan/workspace/grafana/grafana/pkg/server/wire.go:177:20: cycle for *github.com/grafana/grafana/pkg/services/pluginsintegration/plugincontext.Provider:
	*github.com/grafana/grafana/pkg/services/pluginsintegration/plugincontext.Provider (github.com/grafana/grafana/pkg/services/pluginsintegration/plugincontext.ProvideService) ->
	github.com/grafana/grafana/pkg/services/datasources.DataSourceService (github.com/grafana/grafana/pkg/services/datasources/service.ProvideService) ->
```